examples/add/train_add_distill_sd_wds.py

```diff
@@ -0,0 +1,1673 @@
+
+    @property
+    def train_dataset(self):
+        return self._train_dataset
+
+    @property
+    def train_dataloader(self):
+        return self._train_dataloader
+
+
+class Denoiser:
+    def __init__(self, alphas, sigmas, prediction_type="epsilon"):
+        self.alphas = alphas
+        self.sigmas = sigmas
+        self.prediction_type = prediction_type
+
+    def to(self, device):
+        self.alphas = self.alphas.to(device)
+        self.sigmas = self.sigmas.to(device)
+        return self
+
+    def __call__(self, model_output, timesteps, sample):
+        alphas = extract_into_tensor(self.alphas, timesteps, sample.shape)
+        sigmas = extract_into_tensor(self.sigmas, timesteps, sample.shape)
+        if self.prediction_type == "epsilon":
+            pred_x_0 = (sample - sigmas * model_output) / alphas
+        elif self.prediction_type == "sample":
+            pred_x_0 = model_output
+        elif self.prediction_type == "v_prediction":
+            pred_x_0 = alphas * sample - sigmas * model_output
+        else:
+            raise ValueError(
+                f"Prediction type {self.prediction_type} is not supported; currently, `epsilon`, `sample`, and"
+                f" `v_prediction` are supported."
+            )
+
+        return pred_x_0
+
+
+# Based on SpectralConv1d from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L29
+class SpectralConv1d(torch.nn.Conv1d):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        torch.nn.utils.SpectralNorm.apply(self, name="weight", n_power_iterations=1, dim=0, eps=1e-12)
+
+
+# Based on ResidualBlock from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/shared.py#L20
+class ResidualBlock(torch.nn.Module):
+    def __init__(self, fn: Callable):
+        super().__init__()
+        self.fn = fn
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return (self.fn(x) + x) / np.sqrt(2)
+
+
+# Based on make_block from the official StyleGAN-T code
+# https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L64
+class DiscHeadBlock(torch.nn.Module):
+    """
+    StyleGAN-T block: SpectralConv1d => BatchNormLocal => LeakyReLU
+    """
+
+    def __init__(
+        self,
+        channels: int,
+        kernel_size: int,
+        num_groups: int = 8,
+        leaky_relu_neg_slope: float = 0.2,
+    ):
+        super().__init__()
+        self.channels = channels
+
+        self.conv = SpectralConv1d(
+            channels,
+            channels,
+            kernel_size=kernel_size,
+            padding=kernel_size // 2,
+            padding_mode="circular",
+        )
+        self.batch_norm = torch.nn.GroupNorm(num_groups, channels)
```

Comment:
This is group norm no?

Reply:
Just as a note, the original StyleGAN-T implementation uses a class they call [`BatchNormLocal`](https://github.com/autonomousvision/stylegan-t/blob/36ab80ce76237fefe03e65e9b3161c040ae888e3/networks/discriminator.py#L35) and cite [Hoffer, Hubara, and Soudry 2017](https://proceedings.neurips.cc/paper_files/paper/2017/file/a5e0ff62be0b08456fc7f1e88812af3d-Paper.pdf). As far as I can tell the method in that paper and the StyleGAN-T implementation is virtually identical to [group normalization](https://pytorch.org/docs/stable/generated/torch.nn.GroupNorm.html), introduced in [Wu and He 2018](https://openaccess.thecvf.com/content_ECCV_2018/papers/Yuxin_Wu_Group_Normalization_ECCV_2018_paper.pdf). But there might be some small differences (perhaps in how virtual batches/groups are handled across multiple GPUs/nodes).